pkg/servers/provisioner/provisioner.go

```diff
@@ -44,45 +52,144 @@ func New(logger *slog.Logger, client linodeclient.Client) (*Server, error) {
 	return srv, srv.registerMetrics()
 }
 
-// init exists in case someone initializes server with nil logger.
-func (s *Server) init() {
-	if s.log == nil {
-		s.log = slog.Default()
-	}
+// init checks if logger was initialized and starts new span.
+func (s *Server) init(ctx context.Context, caller string) (context.Context, trace.Span) {
+	s.once.Do(func() {
+		if s.log == nil {
+			s.log = slog.Default()
+		}
+	})
+
+	return tracing.Start(ctx, caller)
+}
+
+func (s *Server) logAttr(attr ...slog.Attr) *slog.Logger {
+	return slog.New(s.log.Handler().WithAttrs(attr))
 }
 
 // DriverCreateBucket call is made to create the bucket in the backend.
 //
 // NOTE: this call needs to be idempotent.
 //  1. If a bucket that matches both name and parameters already exists, then OK (success) must be returned.
 //  2. If a bucket by same name, but different parameters is provided, then the appropriate error code ALREADY_EXISTS must be returned.
-func (s *Server) DriverCreateBucket(_ context.Context, _ *cosi.DriverCreateBucketRequest) (*cosi.DriverCreateBucketResponse, error) {
-	s.once.Do(s.init)
-
-	// get bucket
-	// bucket exists:
-	//    compare actual with expected
-	//    return as in 2
-	// bucket not exists:
-	//    create bucket
-	//    return as in 1
-
-	panic("FIXME: unimplemented")
+func (s *Server) DriverCreateBucket(ctx context.Context, req *cosi.DriverCreateBucketRequest) (*cosi.DriverCreateBucketResponse, error) {
+	ctx, span := s.init(ctx, "DriverCreateBucket")
+	defer span.End()
+
+	label := req.GetName()
+	cluster := req.GetParameters()[ParamRegion]
+	cors := ParamCORSValue(req.GetParameters()[ParamCORS])
+
+	acl := linodego.ObjectStorageACL(req.GetParameters()[ParamACL])
+	if acl == "" {
+		acl = linodego.ACLPrivate
+	}
+
+	log := s.logAttr(
+		slog.String(KeyBucketCluster, cluster),
+		slog.String(KeyBucketLabel, label),
+	).WithGroup("DriverCreateBucket")
+
+	span.SetAttributes(
+		attribute.String(KeyBucketCluster, cluster),
+		attribute.String(KeyBucketLabel, label),
+	)
+
+	log.InfoContext(ctx, "bucket creation initiated")
+
+	if cluster == "" {
+		log.ErrorContext(ctx, "required parameter was not provided in the request", "error", ErrMissingRegion)
+
+		return nil, tracing.Error(span, codes.InvalidArgument, ErrMissingRegion)
+	}
+
+	bucket, err := s.client.GetObjectStorageBucket(ctx, cluster, label)
+	if err != nil && !errors.Is(err, ErrNotFound) {
```

Comment:
I think this is quite hard to read:

I'd suggest either a switch or change the construction to something like:

```
if err != nil {
    if !errors.Is(err, ErrNotFound) {
        return early
    }
    
    // Bucket create
}

// Check bucket access
```

I think it's the usage of checking that the error value is not "ErrNotFound" to confirm the bucket exists that's just difficult to comprehend at first glance

Reply:
Fixed it, it still follows the same logic, but now should be more readable:

```mermaid
flowchart TD
    A((Start)) --> B{GetBucket}
    B -- err == BucketNotFound --> C[CreateBucket]
    C --> F((Success))
    B -- error == nil --> E{CompareBuckets}
    B -- error != BucketNotFound --> D((Failure))
    E -- Expected == Actual --> F
    E -- Expected != Actual --> D
```

Reply:
Looks great. Also I <3 mermaid. I use it in obsidian all the time